Moderator Tool Improvements

luap42 opened the pull request:
Some improvements to the moderator tools.
(Documentation comes soon)

ArtOfCode- commented on app/views/users/mod_delete.html.erb:
```diff
@@ -0,0 +1,34 @@
+    </div>
+
+    <div class="widget">
+        <div class="widget--body">
+            <h4>Destroy Acccount</h4>
+            <p>Destroy Accounts of blatant spammers and trolls.</p>
+
+            <%= link_to 'Destroy Account', destroy_user_path(@user.id), remote: true,
+                        method: :delete, class: 'js-destroy-user button is-danger is-filled' %>
+        </div>
```
discuss: We should review whether we actually want this here. Destroying accounts completely is a good way to break the database, and now that we have soft-deletes it's probably not needed.

cellio replied:
I propose to remove "destroy" for the reasons Art said.

cellio replied:
@trichoplax if the destroy option is still there, could you remove it from the UI at least?  Thanks.

cellio replied:
1. Yes, remove the "Destroy Account" button and keep the other three.
2. I defer to @luap42  but I think this is right.
3. I defer to luap42 again but yeah, that sounds better for a public platform rather than our own instance.  I believe "feed to STAT" is delete+ spam-block.

sau226 committed:
Bump co-design version to 0.12.5

Taeir committed:
Fix routes and update links in UI

Taeir committed:
Add 2FA support

Taeir committed:
Apply code suggestions from ArtOfCode

Taeir committed:
Fix mistake in attribute map

Taeir committed:
Update schema

Taeir committed:
Fix warnings for case insensitive uniqueness validations

Taeir committed:
Enable all safe new defaults

Taeir committed:
Fix deprecated where.not usage

Taeir committed:
Fix content_type test

Taeir committed:
Update dependencies

New dependencies require ruby 2.7 and rails 6.0 as minimum versions.

Taeir committed:
Please the linter gods

It seems that the update to rubocop/rails means we get some new rules.
The new rules seem to have changed their minds on how some things should
look or be done. This applies the new rules.

Taeir committed:
Fix docker image

Taeir committed:
Update spring, only load spring in development

Taeir committed:
Fix deprecation warnings for rmagick

Taeir committed:
Check media type rather than content type

Also swap arguments around such that the error message upon failure
lists the right value as expected.

Taeir committed:
Enable other safe defaults (not using sidekiq or resque)

Taeir committed:
Use helpers in devise

Taeir committed:
Add SSO Sign in button on normal sign in page

Taeir committed:
Fix links to SAML

Taeir committed:
Add sign in with sso notice on registrations page

Taeir committed:
Remove comments of safely updated packages

Also add a note in the INSTALLATION about the ruby versions that are
supported/not supported.

Taeir committed:
Update redis library to v5

Taeir committed:
Switch to rails 6 defaults and cleanup safe settings

The only setting left in new_framework_defaults_6.0 is about updating
the cookies. If the system runs stable on rails 6, the file can be
removed and the upgrade is completed.

Taeir committed:
Fix files without .html extension

Rails 6.1 is more picky about the response formats, and not having the
intended html extension can cause issues.

Taeir committed:
Run rails 6.1 update procedure

Taeir committed:
Fix rails 6 image uploading

Taeir committed:
Toggle safe rails 6.1 settings

Taeir committed:
Fix assets

Taeir committed:
Upgrade puma and clean up Gemfile

Package notes are added into the PR instead.

Taeir committed:
Add missing optional mentions on belongs_to relations

Taeir committed:
Add missing translation

Taeir committed:
Fix flag fixture polymorphic relation without type

When a relation is polymorphic, the type of the target must be specified
in the fixture, otherwise the polymorphic_type field is not set and the
record is invalid.

Taeir committed:
Fix categories controller test

Taeir committed:
Don't spam codidact.com during the tests

Taeir committed:
Upgrade to rails 7

Taeir committed:
Update sass-rails to sassc-rails

Fix invalid SCSS

Taeir committed:
Update sass-rails to sassc-rails

Fix invalid SCSS

Taeir committed:
Use sprockets 4

Taeir committed:
Switch to terser for js compression

Taeir committed:
Add schema version

Taeir committed:
Switch to rails 7.0 defaults

Taeir committed:
Fix linter error

Taeir committed:
Fix help posts not editable

The user is not present on those posts, and rails 6 performs better
validations. It must be marked as optional to fix the issue.

Taeir committed:
Ensure saved value is string in redis

Taeir committed:
Replace deprecated method with correct one

Taeir committed:
Fix caching

The namespaced env cache needs to be recreated after forking. With the
old method of changing the Rails cache, this was not happening
correctly.

Taeir committed:
Fix rubocop errors

The namespaced env cache breaks multiple rubocop rules with regards to
using map on a hash-like object. However, the suggested fixes break the
intention of the code. Therefore, I have added it to be ignored.

Taeir committed:
Properly use redis in tests

Taeir committed:
Use fully qualified name for redis

Taeir committed:
Remove new defaults initializer

Otherwise, new checkouts/installs of the project would not run with
recommended settings.

Taeir committed:
Start with migration to rails 6.0

Taeir committed:
Start with migration to rails 6.0

Taeir committed:
Update installation instructions

ArtOfCode- committed:
Update schema for 7

Taeir committed:
Add ruby 3 support

Taeir committed:
Fix CI images

Use less specific versions to prevent image removals from breaking the CI.

Taeir committed:
Update omniauth

Taeir committed:
Add ability to create tags for mods

Closes #180

Taeir committed:
Prevent updating of tag names

ArtOfCode- committed:
Add ability to follow new comment threads on a post; close #819

Taeir committed:
Do not display closed posts in hot posts

Taeir committed:
Add SAML support

- Add devise_saml_authenticatable
- Add saml_authenticatable to user
- Add SamlInit concern + attribute map
- Add devise config for SAML
- Add site settings for SAML
- Add SsoProfile model
- Add conditional SAML routes

cellio committed:
Changed error messages to avoid 'tags can't have more than 5 tags' construct.  I realize that usually we want to attach an error to the field (:tags) not the post (:base), but I couldn't come up with wordings that were clear and the error messages point to tags as the problem, so I think this is ok.

MoshiKoi committed:
Apparently there is actually a save

Taeir committed:
Report redis errors and inherit sockets for development

Taeir committed:
Add missing relation

Taeir committed:
In seeding, actually link to the corresponding objects

Previously, we would pass ids to the creation handler. However, this
does not pass validations whenever the association is non-optional (i.e.
belongs_to). With these changes, we pass the actual Rails objects
instead of just the ids, which does pass the validations and sets it
properly.

Taeir committed:
Add required license field for communities

Taeir committed:
Revert redis to version 4 until issue is fixed and update settings

Settings are now properly loaded from the database.yml config file,
rather than only taking some of those settings.

Taeir committed:
Also hide locked posts

cellio committed:
Generalized wording on vote-summary page so it works for both you and other users.

MoshiKoi committed:
Don't rerender the preview if nothing changed

MoshiKoi committed:
Add net votes for each day to the vote summary

Taeir committed:
Add back still required install instructions

The `rmagick` library requires imagemagick to also be installed (libvips is the new default used by image processing in rails).

Taeir committed:
Fix error in docker

Taeir committed:
Ensure docker works out of the box

Taeir committed:
Add some notes on docker compose and Mac

Also update display

Taeir committed:
Also create storage config in docker

Taeir committed:
Wait until mysql is up properly

Taeir committed:
Remove outdated config option

dependabot[bot] committed:
Bump nokogiri from 1.13.8 to 1.13.9

Bumps [nokogiri](https://github.com/sparklemotion/nokogiri) from 1.13.8 to 1.13.9.
- [Release notes](https://github.com/sparklemotion/nokogiri/releases)
- [Changelog](https://github.com/sparklemotion/nokogiri/blob/main/CHANGELOG.md)
- [Commits](https://github.com/sparklemotion/nokogiri/compare/v1.13.8...v1.13.9)

---
updated-dependencies:
- dependency-name: nokogiri
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump commonmarker from 0.23.5 to 0.23.6

Bumps [commonmarker](https://github.com/gjtorikian/commonmarker) from 0.23.5 to 0.23.6.
- [Release notes](https://github.com/gjtorikian/commonmarker/releases)
- [Changelog](https://github.com/gjtorikian/commonmarker/blob/main/CHANGELOG.md)
- [Commits](https://github.com/gjtorikian/commonmarker/compare/v0.23.5...v0.23.6)

---
updated-dependencies:
- dependency-name: commonmarker
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

ArtOfCode- committed:
Fix upload redirects

ArtOfCode- committed:
Asyncify to fix error

Taeir committed:
Add setting to control 2FA usage of SSO users

Taeir committed:
Disallow SSO users from changing email/password

Taeir committed:
Enforce that SSO users sign-in with SSO

Taeir committed:
Allow disconnect account from SSO when mixed

Taeir committed:
Check for mixed sign in before allowing SSO disconnect

Taeir committed:
Fix impersonation with SSO

Taeir committed:
Consistently use Single Sign-On rather than Single Sign On

dependabot[bot] committed:
Bump loofah from 2.18.0 to 2.19.1

Bumps [loofah](https://github.com/flavorjones/loofah) from 2.18.0 to 2.19.1.
- [Release notes](https://github.com/flavorjones/loofah/releases)
- [Changelog](https://github.com/flavorjones/loofah/blob/main/CHANGELOG.md)
- [Commits](https://github.com/flavorjones/loofah/compare/v2.18.0...v2.19.1)

---
updated-dependencies:
- dependency-name: loofah
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
Bump rails-html-sanitizer from 1.4.3 to 1.4.4

Bumps [rails-html-sanitizer](https://github.com/rails/rails-html-sanitizer) from 1.4.3 to 1.4.4.
- [Release notes](https://github.com/rails/rails-html-sanitizer/releases)
- [Changelog](https://github.com/rails/rails-html-sanitizer/blob/master/CHANGELOG.md)
- [Commits](https://github.com/rails/rails-html-sanitizer/compare/v1.4.3...v1.4.4)

---
updated-dependencies:
- dependency-name: rails-html-sanitizer
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

MoshiKoi committed:
Fix double bracketed footnotes

MoshiKoi committed:
Prevent selection of nonexistent parent tag

cellio committed:
changed order of fields (current pwd first, then new pwd & confirmation)

ArtOfCode- committed:
Remove caching that killed the server

cellio committed:
clarified field labels

MoshiKoi committed:
Move see-all-notifications button to top

MoshiKoi committed:
Add notifications tab to user page

MoshiKoi committed:
Add indicator if there are pending suggested edits

MoshiKoi committed:
Cache whether there are pending suggestions

MoshiKoi committed:
Move pending logic to helper method

Taeir committed:
Fix tour test for the 3rd time

Also add tests to ensure that the tagset is correctly created by the
seeds.

Taeir committed:
Fix CI issues

- Fix rubocop errors accidentally merged
- Fix failing test

Taeir committed:
Apply consistently between sessions_controller and saml version

Taeir committed:
Fix comparisons

Taeir committed:
Remove unnecessary test

trichoplax committed:
Convert 'else if' to 'if' because it depends on data from the previous 'if' block

trichoplax committed:
Remove check for old preferences schema from over 2 years ago

trichoplax committed:
Add early return for most frequent case to reduce processing

trichoplax committed:
Remove null check that is now redundant due to early return

trichoplax committed:
Improve comment to state local variable contains preferences

trichoplax committed:
Reject only null & undefined, rather than all falsy local preferences

trichoplax committed:
Remove 'if' block made redundant by early returns

trichoplax committed:
Store preferences on QPixel rather than window ('this')

trichoplax committed:
Suppress preferences retrieval when user is not signed in

trichoplax committed:
Extract duplicated localStorage preferences code into a function

trichoplax committed:
Use unique localStorage key per user to avoid corruption

trichoplax committed:
Make 'if' explicit and include mandatory braces for code standards

trichoplax committed:
Add the commas I forgot between new functions in QPixel object

trichoplax committed:
Change preference function to use early return

trichoplax committed:
Extract common code into separate functions

trichoplax committed:
Add JavaScript documentation comments for new functions

cellio committed:
mention ruby 3 in installation instructions

cellio committed:
update Ruby version info (since we've moved to Ruby 3 I don't know if we still support 2.7, so hedging a little there)

trichoplax committed:
Fix old internal link that no longer needs a number

trichoplax committed:
Add missing backtick that was hiding a heading from rendered markdown

trichoplax committed:
Replace abandoned heading number with heading wording

trichoplax committed:
Use correct path in link text to avoid confusion

trichoplax committed:
Add missing word and wrap very long line

trichoplax committed:
Replace fake footnotes that unexpectedly linked off page

trichoplax committed:
Remove space preventing italics from working in link text

trichoplax committed:
Replace .com with .org as apitester.com does not exist

trichoplax committed:
Use apitester.org/app to avoid advert for mobile app

trichoplax committed:
Add missing # in internal link

dependabot[bot] committed:
Bump rack from 2.2.4 to 2.2.6.2

Bumps [rack](https://github.com/rack/rack) from 2.2.4 to 2.2.6.2.
- [Release notes](https://github.com/rack/rack/releases)
- [Changelog](https://github.com/rack/rack/blob/main/CHANGELOG.md)
- [Commits](https://github.com/rack/rack/compare/2.2.4...v2.2.6.2)

---
updated-dependencies:
- dependency-name: rack
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

duncanscode committed:
search results count added

Added search results post count and adjusted styling to accommodate

Taeir committed:
Ensure tour tagset works with both strings and ints

dependabot[bot] committed:
Bump globalid from 1.0.0 to 1.0.1

Bumps [globalid](https://github.com/rails/globalid) from 1.0.0 to 1.0.1.
- [Release notes](https://github.com/rails/globalid/releases)
- [Commits](https://github.com/rails/globalid/compare/v1.0.0...v1.0.1)

---
updated-dependencies:
- dependency-name: globalid
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

trichoplax committed:
Add user id and signed in status to HTML body to avoid requests

trichoplax committed:
Check signed in status on HTML body rather than making a request

trichoplax committed:
Access user id from HTML body rather than making a request

duncanscode committed:
Removed cateogry-meta

Adjusted styling to better conform with project standards.

trichoplax committed:
Prevent request for user data unless there is a signed in user

trichoplax committed:
Use HTML data attribute userId instead of redundant signedIn

trichoplax committed:
Update JS doc comment because function is no longer async

trichoplax committed:
Simplify QPixel._preferencesLocalStorageKey function

trichoplax committed:
Move check for signed in user from preferences function to preference

Previously the QPixel.preferences function returned null when there
was no signed in user. This caused an error as the QPixel.preference
function tried to access properties of null. Now the check for a
signed in user happens in QPixel.preference and this function returns
null when there is no signed in user. The three places that currently
use this return value all behave correctly when there is no signed in
user - the null does not cause incorrect behaviour.

trichoplax committed:
Rename function to make clear it is internal only

trichoplax committed:
Cache '_fetchPreferences' returned promise to avoid redundant fetches

MoshiKoi committed:
Update app/models/suggested_edit.rb

Co-authored-by: Taico Aerts <devtaeir@taico.nl>

superplane39 committed:
site -> community

superplane39 committed:
site -> community

cellio committed:
added info about using HTML details/summary (CommonMark doesn't support)

dependabot[bot] committed:
Bump commonmarker from 0.23.6 to 0.23.7

Bumps [commonmarker](https://github.com/gjtorikian/commonmarker) from 0.23.6 to 0.23.7.
- [Release notes](https://github.com/gjtorikian/commonmarker/releases)
- [Changelog](https://github.com/gjtorikian/commonmarker/blob/main/CHANGELOG.md)
- [Commits](https://github.com/gjtorikian/commonmarker/compare/v0.23.6...v0.23.7)

---
updated-dependencies:
- dependency-name: commonmarker
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

ArtOfCode- committed:
Add Stripe setting

ArtOfCode- committed:
Only load Stripe on donation pages

ArtOfCode- committed:
Add all-users email tool

trichoplax committed:
Let only moderators (implicitly including admins) see who handled a flag

trichoplax commented:
Would it be practical to split this pull request into smaller changes to allow some to be deployed and get real world feedback?

I'm wondering because of the issues that mention this pull request. Do they depend on the whole change or just smaller parts of it that might not need to wait for the rest?

cellio commented:
@ArtOfCode-  re the previous comment: do you think it would be easier to subdivide this PR or finish it?  (Asking because you reviewed it in the past, so you have better insights than most other people.)

ArtOfCode- commented:
Probably easier to finish it - would take unnecessary work to split it up. Shouldn't need much more work, though - addressing comments and requests and fixing the conflicts. @trichoplax if you find yourself looking for any more intro to Ruby, this might be doable for you.

trichoplax commented:
@ArtOfCode- Thanks for the suggestion. I'm going to start by trying to rebase `master` onto this branch (in the safety of my own fork), unless anyone tells me that's a bad idea (this will be my first rebase).

trichoplax commented:
I've now rebased `develop` onto my fork's version of this branch, and resolved the conflicts. I won't push the changes back to this branch since I don't have write access so wouldn't be able to continue it here anyway. Instead I'll try to resolve the review comments shown here, but on my fork's version of this branch, and then probably make a new pull request for further review unless someone knows a better way.

cellio commented:
> Instead I'll try to resolve the review comments shown here, but on my fork's version of this branch, and then probably make a new pull request for further review unless someone knows a better way.

Sounds good to me.  We had another PR recently where someone started it and someone else finished, and "new branch + new PR" is what happened there too.  You can see comments here and address them there and it should work.

Congrats on the rebase.  I had no idea how bad the conflicts might be.

Thank you for working on this!

trichoplax commented:
@cellio Thanks! I'll proceed as planned then.

As for the conflicts, there were fewer than I expected. My plan for today was just to make a start on them, but they turned out to be few and fairly easy to resolve. It'll need a careful review from someone to make sure I haven't misinterpreted anything though.

cellio commented:
After you've addressed the outstanding comments here, we can deploy your branch to the dev server for testing in parallel with code review.  That way those of us who aren't qualified to evaluate the code changes can bang on the server and make sure it behaves the way we expect.

trichoplax committed:
Add missing indentation

trichoplax committed:
Fix one line if statement to match Code Standards

trichoplax committed:
Add missing semicolon

trichoplax committed:
Condense two queries into one for efficiency

trichoplax committed:
Switch 'if' branches to prevent leaking information about post existence

trichoplax committed:
Extract private thread checks into a 'before_action'

trichoplax committed:
Remove debug printing code

trichoplax committed:
Remove comma and surplus 's', reduce number of lines

trichoplax committed:
Make method name more explicit

trichoplax committed:
Improve wording and reduce number of lines

trichoplax committed:
Drop commas, improve wording, keep to 140 characters per line

trichoplax committed:
Keep to 140 characters per line

trichoplax committed:
Drop commas and keep to 140 characters per line

trichoplax committed:
Remove redundant usages of 'is_admin' covered by 'is_moderator'

trichoplax committed:
Combine 2 return unless guards into 1 for readability

trichoplax committed:
Remove unnecessary break in line

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax commented:
I've addressed all but the final review comment from this pull request (and added a note about that one as a reminder on the new pull request #1009). I've also changed a few minor things I noticed, and added a few more review comments on the new pull request with questions that occurred to me that I haven't been able to resolve yet.

I'd appreciate anyone's insights over on that pull request.

trichoplax committed:
Add missing closing parenthesis

trichoplax committed:
Update database schema after running database migration for new mod tools

trichoplax committed:
Restrict thread followers to only moderators (including admins)

ArtOfCode- commented:
Superseded by #1009.

luap42 committed:
Ignore rbenv local file

luap42 committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax commented:
@luap42 I noticed you have made some commits to the branch that this pull request was based on, so I wanted to make sure you're up to date on what's happened to this pull request, and why it is currently closed:

- I picked this up and made some changes based on review comments on this pull request, but I couldn't commit the changes in this repo due to lack of access, so I made them in a fork, and opened a new pull request #1009.
- This raised some further questions and more work is required, which I've allowed myself to be distracted from by other things for the last few months.
- In that time some conflicts have arisen that need to be resolved. I expect `develop` needs to be rebased onto this branch again - if I'd realised which branch you were rebasing onto earlier today I would have mentioned this then to save you any further work.

I don't know how much these two branches have now diverged and whether it's worth trying to combine them or better to abandon one or other. I just wanted to write something quickly to let you know before any further work happens in case it affects what you decide to do.

luap42 commented:
@trichoplax thanks for the heads up, today I only did three things on this branch:

- merge your changes into it
- add ".ruby-version" to the gitignore because it is automatically created by rbenv and should not be committed (unless we choose to commit it, which should be an intentional decision not an "oops I just did git add -a")
- merge develop into it

So divergence shouldn't be too much an issue.

I am looking into how much I'll be able to spend some time on this in the near future. I've seen the other MR and will check the reviews there later. Feel free to ping me on Discord for more coordination.

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax commented:
Reopening now that I have write access to this repository and don't need to be working from a fork. I will address any outstanding comments from #1009 here.

trichoplax commented on app/assets/javascripts/application.js:
```diff
@@ -65,6 +65,13 @@ $(document).on('ready', function() {
+
+    evt.target.classList.add("open");
+    evt.stopPropagation();
+  })
```
Applied on my fork and since merged in.

trichoplax commented on app/controllers/comments_controller.rb:
```diff
@@ -56,11 +58,13 @@ def create_thread
-    elsif !@post.can_access?(current_user)
-      return not_found
+    unless @post.nil?
+      if @post.comments_disabled && !current_user.is_moderator && !current_user.is_admin
```
Applied on my fork and since merged in to this branch.

trichoplax commented on app/controllers/comments_controller.rb:
```diff
@@ -94,6 +100,12 @@ def create
+      redirect_to comment_thread_path(@comment_thread.id)
+      return
+    end
+
```
Applied on my fork and since merged in to this branch.

trichoplax commented on app/controllers/mod_warning_controller.rb:
```diff
@@ -44,25 +62,65 @@ def create
+      flash[:error] = 'A network-wide suspension has been applied which may only be lifted ' \
+                      'by global moderators. Community-wide suspensions which have been imposed ' \
+                      'before that global suspensions cannot be lifted at this time (which is ' \
+                      'probably, why you are seeing this error).'
```
Applied on my fork and since merged in to this branch.

trichoplax commented on app/models/mod_warning.rb:
```diff
@@ -2,13 +2,20 @@ class ModWarning < ApplicationRecord
     active && is_suspension && !suspension_end.past?
   end
 
+  def global?
```
Applied on my fork and since merged in to this branch.

trichoplax commented on app/views/mod_warning/new.html.erb:
```diff
@@ -2,78 +2,89 @@
+    <h1>Warn or Suspend User</h1>
+
+    <div class="notice is-warning">
+        <p>Use the warning tool only against users who have violated the site rules. Prefer other measures, such as friendly asking the user to stop certain behaviors in a comment or <%= link_to 'contacting them privately', mod_contact_path(@user) %>.</p>
```
Applied on my fork and since merged in to this branch.

trichoplax commented on app/views/mod_warning/new.html.erb:
```diff
@@ -2,78 +2,89 @@
+        </div>
+        <div class="widget--body">
+            <div class="h-p-4">
+                <p class="is-lead">Review the generated message and add details. Do not add salutations or informations about possible suspensions, as they are generated automatically.</p>
```
Applied on my fork and since merged in to this branch.

trichoplax commented on app/views/mod_warning/new.html.erb:
```diff
@@ -2,78 +2,89 @@
+        </div>
+        <div class="widget--body">
+            <div class="h-p-4">
+                <p class="is-lead">Decide, whether or not to suspend the user, and if, for how long. Choose an optional message shown publicly on the user profile.</p>
```
Applied on my fork and since merged in to this branch.

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Remove 'destroy user' from routes and user interface

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Remove confusing development-only in joke from fail-ban interface

trichoplax committed:
Rename mod tools subsection to avoid Codidact-specific naming

cellio commented:
Where does this work stand?  I see merge conflicts and a comment saying there's one outstanding issue though I'm not sure which.  I resolved some conversations that ended in "I did this" to try to make this a little easier to follow.  It feels like we're 99% there -- what's the remaining 1%?

cellio commented:
When this merges and is deployed, we will be able to answer meta:289200 (user requesting private discussion with mods) by directing people to raise a flag requesting a conversation.  If moderators agree, they can then initiate the private thread in the usual way, since it needn't be tied to a warning.  That should meet the need in this meta request.

trichoplax commented:
@cellio I've just realised I didn't reply here in GitHub. Maybe I mixed this up with my [reply in Discord](https://discord.com/channels/634104110131445811/940041052776177685/1173799889902252153) around the same time.

I haven't found much time to look at this the last month or so. I'm currently putting some time into rebasing onto the `develop` branch, which is resulting in a lot of merge conflicts. This seems like a good sign from a broader perspective (it shows that lots of people have merged other branches since I last worked on this).

cellio commented:
> @cellio I've just realised I didn't reply here in GitHub. 

My comment came from meta triage -- linking another meta question that can be updated when this lands.  I wasn't trying to push you -- but I'm excited to hear you're working on the rebase, as I'd love to see us get this over the finish line somehow.

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Fix post must exist error preventing sending a mod message

trichoplax committed:
Fix error due to lack of post when displaying moderator comment thread

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Update schema.rb following migrations

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Split user mod tools sections and remove Codidact specific wording

trichoplax committed:
Use 'failban' rather than 'hellban' throughout code for consistency

trichoplax committed:
Style moderator tools tab as active on user profile

trichoplax committed:
Make mod tools Full Activity Log compatible with private comment threads

trichoplax committed:
Make mod tools Global Summary compatible with private comment threads

trichoplax committed:
Improve wording of user mod tools page intro message

trichoplax committed:
Make impersonate section of user mod tools show correct side panels

trichoplax committed:
Make user mod tools tab remain highlighted when switching sections

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Include link to full thread for private messages

trichoplax committed:
Indicate when a user has no private messages or warnings

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Remove tests for destroy user functionality that no longer exists

trichoplax committed:
Remove redundant rubocop disabling

trichoplax committed:
Use unless instead of if not thanks to rubocop

trichoplax committed:
Remove redundant whitespace

trichoplax committed:
Remove duplicate of method

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Fix missing mod messages in List Previous section

trichoplax committed:
Remove redundancy from query

trichoplax committed:
Remove redundant safe navigation following explicit nil check

trichoplax committed:
Use parentheses for consistency

trichoplax committed:
Apply Taeir's suggestion from a comment on pull request 1009

Taeir's explanation: NOTE: I use select here to allow rails to merge the two operations into one query and avoid loading the full ThreadFollower objects

trichoplax committed:
Reduce scope of variables no longer used outside method

trichoplax committed:
Reduce confusing variable reuse and mislabelling

trichoplax committed:
Make maximum global suspension 1 year, like local suspensions

trichoplax committed:
Reduce use of .present? for readability

trichoplax committed:
Adjust order and wording of Contact Privately notice

trichoplax committed:
Draw attention to the visibility to moderators of a private message

trichoplax committed:
Allow moderators (and implicitly admins) to view all private messages

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Allow notifications on a thread without a parent post

trichoplax committed:
Show username at top of private thread to avoid confusion

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

cellio committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

Oaphi committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

Oaphi committed:
rubocop fix

cellio commented:
I was going to update this from `develop` but apparently there are conflicts now, so that'll require some human attention.

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

cellio commented:
We're starting to add some other small improvements for mods, and I don't want to conflict with work here that's coming, but also this PR is old and I don't think @luap42 wrote that documentation mentioned in the description and I've lost track of what's here.  What's the easiest way to get an overview of what's here?

trichoplax committed:
Merge branch 'develop' into luap42/moderator-tools-improvements

trichoplax commented:
> We're starting to add some other small improvements for mods, and I don't want to conflict with work here that's coming, but also this PR is old and I don't think @luap42 wrote that documentation mentioned in the description and I've lost track of what's here. What's the easiest way to get an overview of what's here?

Sorry for taking 2 and a half weeks to reply to this. I'm back now and have brought this branch up to date with changes to the `develop` branch. The only conflict that required a change was for the `(n messages)` addition to the Moderator Tools link on the user profile page. This link no longer exists in this branch, instead being replaced with an additional tab on the user page. I've therefore added the `(n messages)` to the title of that tab. The screenshots below show the difference. This avoids losing that change altogether, and it can be further improved later if there are any suggestions.

On `develop` branch:

![Screenshot 2025-03-01 at 18-23-08 DEV QPixel - User user](https://github.com/user-attachments/assets/e5303a75-6264-46d8-97dd-471ce2556994)

On this branch:

![Screenshot 2025-03-01 at 18-23-45 DEV QPixel - User user](https://github.com/user-attachments/assets/dc863d02-ae6f-45b0-b867-47b8aa10cadc)

## Overview of changes in this branch
### Mod tools tab
The "Moderator Tools: user" page is replaced with a "Moderator Tools" tab on the user page.

On `develop` branch:

![Screenshot 2025-03-01 at 18-43-05 DEV QPixel - Moderator Tools user](https://github.com/user-attachments/assets/0c3ff9f1-111b-4946-8209-9e934daf3f7b)

On this branch:

![Screenshot 2025-03-01 at 18-43-45 DEV QPixel - Moderator Tools user](https://github.com/user-attachments/assets/7382dfa2-4545-4be8-a709-7342a3d2fa72)

### Private messages
A moderator can now send a private message to a user using the comments system, without it needing to be a warning or suspension:

![Screenshot 2025-03-01 at 18-46-57 DEV QPixel - Moderator Tools user](https://github.com/user-attachments/assets/8353fa62-e2ec-451a-bb7e-43c05ad44124)

cellio commented:
@trichoplax thanks very much for the screenshots and overview!  And sorry for inadvertently creating that conflict on the count; I completely forgot how this was arranged on this branch.  I'm really looking forward to having this on main, especially the private conversations.

trichoplax commented:
@cellio Please don't apologise for adding useful new features. It's been great to see so much improving recently. Conflicts are a natural part of that. If there's a particular potential conflict that occurs to you when working on a new feature feel free to ask, but I'd prefer that you be able to just focus on your own changes - that's the beauty of branches. Now that I'm back I'll be glancing over other pull requests and mentioning anything that might be worth discussion while this one is being finalised.

cellio commented:
For purposes of (a) cataloguing and (b) possibly splitting off parts of the work to merge sooner, here's at least a partial list of the functional changes on this branch.  (We didn't write doc back then, and it's been a while since we had this set up on a test server for poking, so my memory is imperfect.  Edits not only welcome but encouraged!)

[ ] https://github.com/codidact/qpixel/issues/584 (mod-private headless comment threads)
[ ] https://github.com/codidact/qpixel/issues/757 
[ ] https://github.com/codidact/qpixel/issues/766 (incomplete? I think you can reset but it's not logged?)
[ ] new mod-tools pages: replace the menu with the sidebar seen in https://github.com/codidact/qpixel/pull/767#discussion_r1298572207 and accessed like in the screenshot in https://github.com/codidact/qpixel/pull/767#issuecomment-2692364107 (and adjust some section views) -- what does this layout do on a phone?
[ ] user dashboard page like in the screenshot in https://github.com/codidact/qpixel/pull/767#issuecomment-2692364107
[ ] global mod tools: summary (what's in here?), network suspension
[ ] mod tools: "contact community team" button

Future work that this PR enables, but not to be done here (collected these links from throughout the PR):

[ ] https://github.com/codidact/qpixel/issues/793
[ ] https://github.com/codidact/qpixel/issues/749
[ ] https://github.com/codidact/qpixel/issues/903
[ ] https://github.com/codidact/qpixel/issues/996
[ ] https://github.com/codidact/qpixel/issues/1390

cellio commented:
For purposes of (a) cataloguing and (b) possibly splitting off parts of the work to merge sooner, here's at least a partial list of the functional changes on this branch.  (We didn't write doc back then, and it's been a while since we had this set up on a test server for poking, so my memory is imperfect.  Edits not only welcome but encouraged!)

-  [ ] (high priority) https://github.com/codidact/qpixel/issues/584 (mod-private headless comment threads)
-  [ ] (high priority) https://github.com/codidact/qpixel/issues/757 
-  [ ] https://github.com/codidact/qpixel/issues/766 (incomplete? I think you can reset but it's not logged?)
-  [x] new mod-tools pages: replace the menu with the sidebar seen in https://github.com/codidact/qpixel/pull/767#discussion_r1298572207 and accessed like in the screenshot in https://github.com/codidact/qpixel/pull/767#issuecomment-2692364107 (and adjust some section views) -- what does this layout do on a phone?
-  [x] user dashboard page like in the screenshot in https://github.com/codidact/qpixel/pull/767#issuecomment-2692364107
-  [ ] global mod tools: summary (what's in here?), network suspension
-  [ ] mod tools: "contact community team" button

Future work that this PR enables, but not to be done here (collected these links from throughout the PR):

-  [ ] https://github.com/codidact/qpixel/issues/793  
-  [ ] https://github.com/codidact/qpixel/issues/749
-  [ ] https://github.com/codidact/qpixel/issues/903
-  [ ] https://github.com/codidact/qpixel/issues/996
-  [ ] https://github.com/codidact/qpixel/issues/1390
